onnxruntime/python/tools/kernel_explorer/kernels/gemm_ck.cc

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "python/tools/kernel_explorer/kernels/gemm_ck.h"
+
+#include <pybind11/stl.h>
+
+#include <utility>
+#include <vector>
+#include <memory>
+
+#include "ck/ck.hpp"
+#include "ck/tensor_operation/gpu/device/tensor_layout.hpp"
+#include "ck/tensor_operation/gpu/device/device_gemm.hpp"
+#include "ck/tensor_operation/gpu/element/element_wise_operation.hpp"
+#include "ck/library/tensor_operation_instance/gpu/gemm.hpp"
+#include "python/tools/kernel_explorer/kernels/gemm.h"
+
+namespace py = pybind11;
+
+namespace onnxruntime {
+
+namespace {
+
+template <typename T>
+struct DataTypeAdaptor {
+  using type = T;
+};
+
+template <>
+struct DataTypeAdaptor<half> {
+  using type = ck::half_t;
+};
+
+using Row = ck::tensor_layout::gemm::RowMajor;
+using Col = ck::tensor_layout::gemm::ColumnMajor;
+
+}  // namespace
+
+template <typename T>
+class CKGemm : public GemmBase<T> {
+  using CKDataType = typename DataTypeAdaptor<T>::type;
+
+  template <typename ALayout, typename BLayout>
+  using DeviceGemm = ck::tensor_operation::device::DeviceGemm<
+      ALayout, BLayout, Row,
+      CKDataType, CKDataType, CKDataType,
+      ck::tensor_operation::element_wise::PassThrough,
+      ck::tensor_operation::element_wise::PassThrough,
+      ck::tensor_operation::element_wise::PassThrough>;
+
+  template <typename Op>
+  using CKGemmInstanceFactory = ck::tensor_operation::device::instance::DeviceOperationInstanceFactory<Op>;
+
+  template <typename Op>
+  static std::vector<std::unique_ptr<ck::tensor_operation::device::BaseOperator>> GetInstances() {
+    auto ptrs = CKGemmInstanceFactory<Op>::GetInstances();
```

Comment:
Can we just do this: return ptrs; ?

If so, maybe we don't need this function at all.

Reply:
Not really. The `CKGemmInstanceFactory<Op>::GetInstances()` returns a `vector<DeviceGemm<...>>`. There is no automatic coercion to `vector<BaseOperator>`. I think this should be reported to ck to 'fix' it.

Reply:
I see. Add this as a comment (just to make it a bit easier for the reader, in case the reader is wondering about the functionality of this seemingly useless function)?